hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -623,6 +630,118 @@ struct TypeVisitor : OptimizerVisitor, visitor::PreOrder<bool, TypeVisitor> {
     }
 };
 
+struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, ConstantFoldingVisitor> {
+    std::map<uint64_t, bool> _constants;
+
+    void collect(Node& node) override {
```

Comment:
Could we move the collect/prune methods into the base class, or are just diverting too much between different optimizer visitors?

Reply:
I thought so too initially, but this is tricky. These methods invoke `walk` which comes directly from `Visitor` and can only invoked from something which knows all the matchers the final visitor has. Using straight-forward CRTP for `OptimizerVisitor` makes it impossible to have a container of different visitors below.

I'd prefer to leave this as is for now and possible clean this up later.